File: leetcode/binary_search/3231_Minimum_Number_of_Increasing_Subsequence_to_Be_Removed/codes/bs.c
#include <stdlib.h>

int upper_bound(int* arr, int size, int target) {
    int lo = 0, hi = size;
    while (lo < hi) {
        int mid = (lo + hi) / 2;
        if (arr[mid] <= target) lo = mid + 1;
        else hi = mid;
    }
    return lo;
}

int minOperations(int* nums, int numsSize) {
    int* tails = (int*)malloc(sizeof(int) * numsSize);
    int size = 0;
    for (int i = 0; i < numsSize; ++i) {
        int neg = -nums[i];
        int idx = upper_bound(tails, size, neg);
        if (idx == size) tails[size++] = neg;
        else tails[idx] = neg;
    }
    free(tails);
    return size;
}

File: leetcode/binary_search/3231_Minimum_Number_of_Increasing_Subsequence_to_Be_Removed/codes/bs.py
from typing import List
import bisect

class Solution:
    def minOperations(self, nums: List[int]) -> int:
        tails = []
        for x in nums:
            i = bisect.bisect_right(tails, -x)
            if i == len(tails):
                tails.append(-x)
            else:
                tails[i] = -x
        return len(tails)
